diff --git a/libnotify/notification.c b/libnotify/notification.c
index 3ca0982..35c6d86 100644
--- a/libnotify/notification.c
+++ b/libnotify/notification.c
@@ -278,7 +278,7 @@ notify_notification_set_property (GObject      *object,
                                   guint         prop_id,
                                   const GValue *value,
                                   GParamSpec   *pspec)
-{
+{  
         NotifyNotification        *notification = NOTIFY_NOTIFICATION (object);
         NotifyNotificationPrivate *priv = notification->priv;
 
@@ -1093,22 +1093,6 @@ notify_notification_show (NotifyNotification *notification,
         }
 
         priv = notification->priv;
-        proxy = _notify_get_proxy (error);
-        if (proxy == NULL) {
-                return FALSE;
-        }
-
-        if (priv->proxy_signal_handler == 0) {
-                priv->proxy_signal_handler = g_signal_connect_object (proxy,
-                                                                      "g-signal",
-                                                                      G_CALLBACK (proxy_g_signal_cb),
-                                                                      notification,
-                                                                      0);
-        }
-
-        if (_notify_uses_portal_notifications ()) {
-                return add_portal_notification (proxy, notification, error);
-        }
 
         g_variant_builder_init (&actions_builder, G_VARIANT_TYPE ("as"));
         for (l = priv->actions; l != NULL; l = l->next) {
@@ -1155,23 +1139,56 @@ notify_notification_show (NotifyNotification *notification,
                                        g_variant_new_string (application_id));
             }
         }
-
-        /* TODO: make this nonblocking */
-        result = g_dbus_proxy_call_sync (proxy,
-                                         "Notify",
-                                         g_variant_new ("(susssasa{sv}i)",
-                                                        priv->app_name ? priv->app_name : notify_get_app_name (),
-                                                        priv->id,
-                                                        priv->icon_name ? priv->icon_name : "",
-                                                        priv->summary ? priv->summary : "",
-                                                        priv->body ? priv->body : "",
-                                                        &actions_builder,
-                                                        &hints_builder,
-                                                        priv->timeout),
-                                         G_DBUS_CALL_FLAGS_NONE,
-                                         -1 /* FIXME ? */,
-                                         NULL,
-                                         error);
+        
+    GError *error2 = NULL;
+    /* Connexion to Dbus session */
+    GDBusConnection *connection;
+    connection = g_bus_get_sync(G_BUS_TYPE_SESSION, NULL, &error2);
+    if (!connection) {
+        g_printerr("Erreur connexion DBus: %s\n", error2->message);
+        g_error_free(error2);
+        return 1;
+    }
+
+    /* Method arguments */
+    const gchar *channel = "whatslectron.pparent_whatslectron";
+    const gchar json_payload[500];
+    g_snprintf(json_payload,500, "{\"message\": \"%s\", "
+        "\"notification\": {"
+            "\"card\": {"
+                "\"summary\": \"%s\", "
+                "\"body\": \"%s\", "
+                "\"icon\": \"/opt/click.ubuntu.com/whatslectron.pparent/current/icon.png\","                
+                "\"popup\": true, "
+                "\"persist\": true"
+            "},"
+            "\"sound\": true,"
+            "\"vibrate\": true"
+        "}"
+        "}",priv->summary,priv->summary,priv->body) ;
+
+    /* DBus call to lomiri notifications (Postal service) */
+    result = g_dbus_connection_call_sync(
+        connection,
+        "com.lomiri.Postal",                          /* dest */
+        "/com/lomiri/Postal/whatslectron_2epparent",         /* object path */
+        "com.lomiri.Postal",                     /* interface */
+        "Post",                                       /* méthode */
+        g_variant_new("(ss)", channel, json_payload), /* paramètres */
+        NULL,                                         /* type de retour */
+        G_DBUS_CALL_FLAGS_NONE,
+        -1,
+        NULL,
+        &error2
+    );
+
+
+   if (!result) {
+        g_printerr("Error in DBus call: %s\n", error2->message);
+        g_error_free(error2);
+        g_object_unref(connection);
+        return 1;
+    }
         if (result == NULL) {
                 return FALSE;
         }
