diff --git a/gtk-input-context/client-gtk/client-imcontext-gtk.c b/gtk-input-context/client-gtk/client-imcontext-gtk.c
index e26ce37..641ba9e 100644
--- a/gtk-input-context/client-gtk/client-imcontext-gtk.c
+++ b/gtk-input-context/client-gtk/client-imcontext-gtk.c
@@ -32,6 +32,61 @@
 #include "client-imcontext-gtk.h"
 #include "qt-gtk-translate.h"
 #include "debug.h"
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <stdio.h>
+
+void xdotool_hack_unlock_keyboard() {
+    DBG("Calling xdotool Hack")
+    pid_t pid = fork();
+    if (pid < 0) {
+        DBG("fork failed");
+        return;
+    }
+
+    if (pid == 0) {
+        // enfant : remplacer le processus par xdotool
+        execlp("xdotool", "xdotool", "key", "Shift_L", (char *)NULL);
+
+        // si exec échoue
+        DBG("execlp failed");
+        _exit(1);
+    } else {
+        // parent : optionnel, attendre que l'enfant se termine
+        int status;
+        waitpid(pid, &status, 0);
+    }
+}
+
+void xdotool_type(const char *text) {
+    DBG("Calling xdotool Hack")
+    pid_t pid = fork();
+    if (pid < 0) {
+        DBG("fork failed");
+        return;
+    }
+
+    if (pid == 0) {
+        // enfant : remplacer le processus par xdotool
+        execlp("xdotool", "xdotool", "type", text, (char *)NULL);
+
+        // si exec échoue
+        DBG("execlp failed");
+        _exit(1);
+    } else {
+        // parent : optionnel, attendre que l'enfant se termine
+        int status;
+        waitpid(pid, &status, 0);
+    }
+}
+
+static gint64 ts_last_back_rst = 0;
+static int needToReset=0;
+static int needToPreventBackspace=0;
+static gboolean getstring_called=0;
+static gboolean nopreedit_detected = FALSE;
 
 static GType _maliit_im_context_type = 0;
 static GtkIMContextClass *parent_class = NULL;
@@ -339,6 +394,8 @@ maliit_im_context_init(MaliitIMContext *
 static void
 maliit_im_context_focus_in(GtkIMContext *context)
 {
+    getstring_called=FALSE;
+    nopreedit_detected=FALSE;
     MaliitIMContext *im_context = MALIIT_IM_CONTEXT(context);
     gboolean focus_changed = TRUE;
     GError *error = NULL;
@@ -378,6 +435,8 @@ maliit_im_context_focus_in(GtkIMContext
         g_clear_error(&error);
     }
 
+    if ( ! im_context->client_window )
+        xdotool_hack_unlock_keyboard();
     // TODO: anything else than call "activateContext" and "showInputMethod" ?
 }
 
@@ -494,6 +553,8 @@ maliit_im_context_reset(GtkIMContext *co
 {
     MaliitIMContext *im_context = MALIIT_IM_CONTEXT(context);
     GError *error = NULL;
+    
+    needToReset=0;
 
     if (!maliit_is_running())
         return;
@@ -524,6 +585,7 @@ maliit_im_context_reset(GtkIMContext *co
 static void
 maliit_im_context_get_preedit_string(GtkIMContext *context, gchar **str, PangoAttrList **attrs, gint *cursor_pos)
 {
+    getstring_called=TRUE;
     MaliitIMContext *im_context = MALIIT_IM_CONTEXT(context);
 
     if (!maliit_is_running()) {
@@ -712,6 +774,34 @@ maliit_im_context_im_initiated_hide(Mali
     return FALSE;
 }
 
+gboolean preedit_timer_callback(gpointer data)
+{
+    MaliitIMContext *context = MALIIT_IM_CONTEXT(GTK_IM_CONTEXT(data));
+    if (getstring_called == FALSE)
+    {
+            xdotool_type(context->preedit_str);
+            g_free(context->preedit_str);
+            context->preedit_str = g_strdup("");
+            context->preedit_cursor_pos = 0;
+            nopreedit_detected = TRUE;
+            maliit_im_context_reset(GTK_IM_CONTEXT(context));
+    }
+
+    return G_SOURCE_REMOVE;  // équivalent à FALSE
+}
+
+static guint reset_watchdog_id = 0;
+gboolean reset_timer_callback(gpointer data)
+{
+    MaliitIMContext *context = MALIIT_IM_CONTEXT(GTK_IM_CONTEXT(data));
+    DBG("RESET TIMER");
+    g_free(context->preedit_str);
+    context->preedit_str = g_strdup("");
+    context->preedit_cursor_pos = 0;    
+    maliit_im_context_reset(GTK_IM_CONTEXT(context)); 
+    return G_SOURCE_REMOVE;  // équivalent à FALSE
+}
+
 gboolean
 maliit_im_context_commit_string(MaliitContext *obj,
                               GDBusMethodInvocation *invocation,
@@ -724,6 +814,20 @@ maliit_im_context_commit_string(MaliitCo
     DBG("string is:%s", string);
 
     MaliitIMContext *im_context = MALIIT_IM_CONTEXT(user_data);
+                
+    if ( nopreedit_detected == TRUE )
+    {
+        xdotool_type(string);
+        g_free(im_context->preedit_str);
+        im_context->preedit_str = g_strdup("");
+        im_context->preedit_cursor_pos = 0;
+        nopreedit_detected = TRUE;
+        maliit_im_context_reset(GTK_IM_CONTEXT(im_context));
+        return TRUE;
+    }
+    else if ( getstring_called == FALSE )
+        g_timeout_add(200, preedit_timer_callback,im_context);
+    
     if (im_context != focused_im_context)
         return FALSE;
 
@@ -734,9 +838,16 @@ maliit_im_context_commit_string(MaliitCo
         g_signal_emit_by_name(focused_im_context, "preedit-changed");
         g_signal_emit_by_name(focused_im_context, "commit", string);
         maliit_context_complete_commit_string(obj, invocation);
+        if (string && (! *string || (g_unichar_isalpha(g_utf8_get_char(g_utf8_find_prev_char(string, string + strlen(string)))) || g_unichar_isspace(g_utf8_get_char(g_utf8_find_prev_char(string, string + strlen(string)))) ) ) )
+             maliit_im_context_reset(GTK_IM_CONTEXT(im_context));
+        else
+        {
+            needToReset=1;
+            needToPreventBackspace=1;
+        }
         return TRUE;
     }
-
+    
     return FALSE;
 }
 
@@ -799,6 +910,24 @@ maliit_im_context_update_preedit(MaliitC
     DBG("im_context = %p string = %s cursorPos = %d", im_context, string, cursorPos);
 
     if (focused_im_context) {
+        
+         if (strlen(string)>0 && ( !im_context->preedit_str || strlen(im_context->preedit_str)==0 ) && ( strlen(string)>1 || g_get_real_time()-ts_last_back_rst< 350000 ) )
+         {
+           return TRUE;
+         }
+        
+        if ( nopreedit_detected == TRUE )
+        {
+        xdotool_type(string);
+        g_free(im_context->preedit_str);
+        im_context->preedit_str = g_strdup("");
+        im_context->preedit_cursor_pos = 0;
+        nopreedit_detected = TRUE;
+        maliit_im_context_reset(GTK_IM_CONTEXT(im_context));
+        return TRUE;
+        }
+        else if ( getstring_called == FALSE )
+            g_timeout_add(200, preedit_timer_callback,im_context);
         guint iter;
         PangoAttrList* attrs;
 
@@ -896,11 +1025,34 @@ maliit_im_context_key_event(MaliitContex
 
     if (focused_im_context)
         window = focused_im_context->client_window;
+    
+    if ( !window )
+    {
+        xdotool_hack_unlock_keyboard();
+    }
 
     event = qt_key_event_to_gdk(type, key, modifiers, text, window);
     if (!event)
         return FALSE;
 
+    if ( event->type == GDK_KEY_PRESS && event->keyval == XK_BackSpace ) 
+    {
+        if (reset_watchdog_id != 0) 
+            g_source_remove(reset_watchdog_id);
+    
+        if (needToReset)
+            reset_watchdog_id = g_timeout_add(500, reset_timer_callback, im_context);
+        
+        if ( g_get_real_time()-ts_last_back_rst< 250000 )
+            return FALSE;
+        
+        if ( needToPreventBackspace==1 )
+        {
+        needToPreventBackspace=0;
+        ts_last_back_rst=g_get_real_time();
+        }
+    }
+
     event->send_event = TRUE;
     event->state |= IM_FORWARD_MASK;
