diff --git a/gtk-input-context/client-gtk/client-imcontext-gtk.c b/gtk-input-context/client-gtk/client-imcontext-gtk.c
index e26ce37..641ba9e 100644
--- a/gtk-input-context/client-gtk/client-imcontext-gtk.c
+++ b/gtk-input-context/client-gtk/client-imcontext-gtk.c
@@ -32,6 +32,59 @@
 #include "client-imcontext-gtk.h"
 #include "qt-gtk-translate.h"
 #include "debug.h"
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <stdio.h>
+
+void xdotool_hack_unlock_keyboard() {
+    DBG("Calling xdotool Hack")
+    pid_t pid = fork();
+    if (pid < 0) {
+        DBG("fork failed");
+        return;
+    }
+
+    if (pid == 0) {
+        // enfant : remplacer le processus par xdotool
+        execlp("xdotool", "xdotool", "key", "Shift_L", (char *)NULL);
+
+        // si exec échoue
+        DBG("execlp failed");
+        _exit(1);
+    } else {
+        // parent : optionnel, attendre que l'enfant se termine
+        int status;
+        waitpid(pid, &status, 0);
+    }
+}
+
+void xdotool_type(const char *text) {
+    DBG("Calling xdotool Hack")
+    pid_t pid = fork();
+    if (pid < 0) {
+        DBG("fork failed");
+        return;
+    }
+
+    if (pid == 0) {
+        // enfant : remplacer le processus par xdotool
+        execlp("xdotool", "xdotool", "type", text, (char *)NULL);
+
+        // si exec échoue
+        DBG("execlp failed");
+        _exit(1);
+    } else {
+        // parent : optionnel, attendre que l'enfant se termine
+        int status;
+        waitpid(pid, &status, 0);
+    }
+}
+
+
+static gboolean getstring_called=0;
+static gboolean nopreedit_detected = FALSE;
 
 static GType _maliit_im_context_type = 0;
 static GtkIMContextClass *parent_class = NULL;
@@ -339,6 +392,8 @@
 static void
 maliit_im_context_focus_in(GtkIMContext *context)
 {
+    getstring_called=FALSE;
+    nopreedit_detected=FALSE;
     MaliitIMContext *im_context = MALIIT_IM_CONTEXT(context);
     gboolean focus_changed = TRUE;
     GError *error = NULL;
@@ -378,6 +433,8 @@
         g_clear_error(&error);
     }
 
+    if ( ! im_context->client_window )
+        xdotool_hack_unlock_keyboard();
     // TODO: anything else than call "activateContext" and "showInputMethod" ?
 }
 
@@ -524,6 +581,7 @@
 static void
 maliit_im_context_get_preedit_string(GtkIMContext *context, gchar **str, PangoAttrList **attrs, gint *cursor_pos)
 {
+    getstring_called=TRUE;
     MaliitIMContext *im_context = MALIIT_IM_CONTEXT(context);
 
     if (!maliit_is_running()) {
@@ -712,6 +770,23 @@
     return FALSE;
 }
 
+gboolean preedit_timer_callback(gpointer data)
+{
+    MaliitIMContext *context = MALIIT_IM_CONTEXT(GTK_IM_CONTEXT(data));
+    if (getstring_called == FALSE)
+    {
+            xdotool_type(context->preedit_str);
+            g_free(context->preedit_str);
+            context->preedit_str = g_strdup("");
+            context->preedit_cursor_pos = 0;
+            nopreedit_detected = TRUE;
+            maliit_im_context_reset(GTK_IM_CONTEXT(context));
+    }
+
+    return G_SOURCE_REMOVE;  // équivalent à FALSE
+}
+
+
 gboolean
 maliit_im_context_commit_string(MaliitContext *obj,
                               GDBusMethodInvocation *invocation,
@@ -724,6 +799,20 @@
     DBG("string is:%s", string);
 
     MaliitIMContext *im_context = MALIIT_IM_CONTEXT(user_data);
+                
+    if ( nopreedit_detected == TRUE )
+    {
+        xdotool_type(string);
+        g_free(im_context->preedit_str);
+        im_context->preedit_str = g_strdup("");
+        im_context->preedit_cursor_pos = 0;
+        nopreedit_detected = TRUE;
+        maliit_im_context_reset(GTK_IM_CONTEXT(im_context));
+        return TRUE;
+    }
+    else if ( getstring_called == FALSE )
+        g_timeout_add(200, preedit_timer_callback,im_context);
+    
     if (im_context != focused_im_context)
         return FALSE;
 
@@ -799,6 +888,38 @@
     DBG("im_context = %p string = %s cursorPos = %d", im_context, string, cursorPos);
 
     if (focused_im_context) {
+        
+         if (strlen(string)>1 && strlen(im_context->preedit_str)==0)
+         {
+            while (gdk_events_pending()) {
+            gtk_main_iteration_do(FALSE);
+                }
+         }
+        if (strlen(string)==0)
+        {
+            g_free(im_context->preedit_str);
+            im_context->preedit_str = g_strdup("");
+            im_context->preedit_cursor_pos = 0;
+            while (gdk_events_pending()) {
+            gtk_main_iteration_do(FALSE);
+                }            
+            maliit_im_context_reset(im_context);
+        }
+        
+        DBG("STEP2 !!! im_context = %p string = %s cursorPos = %d", im_context, string, cursorPos);
+        
+        if ( nopreedit_detected == TRUE )
+        {
+        xdotool_type(string);
+        g_free(im_context->preedit_str);
+        im_context->preedit_str = g_strdup("");
+        im_context->preedit_cursor_pos = 0;
+        nopreedit_detected = TRUE;
+        maliit_im_context_reset(GTK_IM_CONTEXT(im_context));
+        return TRUE;
+        }
+        else if ( getstring_called == FALSE )
+            g_timeout_add(200, preedit_timer_callback,im_context);
         guint iter;
         PangoAttrList* attrs;
 
@@ -896,13 +1017,18 @@
 
     if (focused_im_context)
         window = focused_im_context->client_window;
+    
+    if ( !window )
+    {
+        xdotool_hack_unlock_keyboard();
+    }
 
     event = qt_key_event_to_gdk(type, key, modifiers, text, window);
     if (!event)
         return FALSE;
 
     event->send_event = TRUE;
-    event->state |= IM_FORWARD_MASK;
+    event->state |= IM_FORWARD_MASK;  
 
     gdk_event_put((GdkEvent *)event);
     gdk_event_free((GdkEvent *)event);
